### PR TITLE
feat(treesitter): add support for wasm parsers

### PR DESCRIPTION
**Pull request by lewis6991**
### Problem

Treesitter parsers are currently distributed as compiled shared objects which need to be compiled by the users (with the help of nvim-treesitter).

### Solution

Allow loading wasm built parsers.

Notes:
- For now this is an optional build-time feature (via `ENABLE_WASMTIME`) which will be dynamically linked.
  - Builds with `make CMAKE_EXTRA_FLAGS=-DENABLE_WASMTIME=ON DEPS_CMAKE_FLAGS=-DENABLE_WASMTIME=ON`
- Treesitter requires being built with `TREE_SITTER_FEATURE_WASM` and also requires `wasmtime` at build time. So no `[uv_]dlopen` 😢 
- `wasmtime` has includes that require C11 (`static_assert`).
- To reduce re-compilation times, install `sccache` and build with `RUSTC_WRAPPER=sccache make ...`.

### Status

Waiting for the next tree-sitter release with the commit we're pulling in here to allow us to build against a wasmtime release with the build changes required to avoid unacceptable regressions on macOS. (Ideally, that release would also be built against the same version we use, but that is blocked by the same build system changes; luckily so far the versions have not become incompatible for our purpose.)

### CODE REVIEW

File: cmake.deps/cmake/BuildWasmtime.cmake
Comment (clason) [inline cmake.deps/cmake/BuildWasmtime.cmake:4]:
```suggestion
  GIT_TAG release-19.0.2
```
might as well use the latest and greatest ;) (though 20.0.0 is scheduled for today or tomorrow)

Comment (dundargoc) [reply cmake.deps/cmake/BuildWasmtime.cmake:4]: Derp, this was old code. Should be removed, thanks.

File: .github/scripts/install_deps.sh
Comment (dundargoc) [inline .github/scripts/install_deps.sh:40]: Delete

File: cmake.deps/cmake/TreesitterCMakeLists.txt
Comment (lewis6991) [inline cmake.deps/cmake/TreesitterCMakeLists.txt:14]: @dundargoc is this the right way to do it?

Comment (dundargoc) [reply cmake.deps/cmake/TreesitterCMakeLists.txt:14]: This is a bit sus. Why is this needed? It should in theory work without this.

Comment (dundargoc) [reply cmake.deps/cmake/TreesitterCMakeLists.txt:14]: Aha, I understand now. Give me a sec.

File: cmake/Deps.cmake
Comment (dundargoc) [inline cmake/Deps.cmake:11]: TODO(dundargoc): investigate if we should nuke c99 across the board

File: cmake/FindWasmtime.cmake
Comment (dundargoc) [inline cmake/FindWasmtime.cmake:2]: TODO(dundargoc): hacky, should add a directory for the `.deps` directory directly, alternatively ensure the relevant files are in usr.

File: cmake.deps/deps.txt
Comment (dundargoc) [inline cmake.deps/deps.txt:63]: TODO(dundargoc): this some bullshit, wasmtime don't have the header ´wasm.h´ in their release branches. This means we can't use any of their release branches or release tags with the current setup. Unsure if using a main commit is good for now or we need to ask wasmtime why they're doing this.

File: cmake.deps/cmake/BuildWasmtime.cmake
Comment (dundargoc) [inline cmake.deps/cmake/BuildWasmtime.cmake:5]: TODO(dundargoc): this will always use release which isn't great. Could use a genex here to fix it. Alternatively we could fix the cmake file in wasmtime and use that instead.

**Comment by clason** [comment]
Is this something we want to add to `BUILD.md` (yet)?

**Comment by lewis6991** [comment]
Maybe, do you have anything to propose? I'm not sure what to add.

**Comment by clason** [comment]
Not yet; I agree the wording needs to be careful. We want to phase this in slowly.

**Comment by clason** [comment]
@lewis6991 wasmtime 20.0.0 just got released (and is already updated on homebrew). Could you test your PR again to see if this is breaking?

**Comment by dundargoc** [comment]
We are not using the release tags, where are using random commit on main. We can bump the commit version of course.

**Comment by clason** [comment]
Ah, right. Yes, that would be interesting (to see _how_ breaking the wasmtime update velocity is for us). 

But can't we use release tags (not releases)? I think that would be more meaningful and easier to handle than random commits.

**Comment by dundargoc** [comment]
> But can't we use release tags (not releases)? 

Nah, I tried. `wasm.h` header isn't included in the release tags for some reason, which is needed. `wasm.h` is from https://github.com/WebAssembly/wasm-c-api/blob/main/include/wasm.h that they have vendored on the main branch.

I can ask in wasmtime repo how we're meant to use a release tag since `wasm.h` isn't included in their release tags.

**Comment by clason** [comment]
I just checked; https://github.com/bytecodealliance/wasmtime/archive/v20.0.0.tar.gz _does_ contain `wasm.h`?

File: cmake.deps/deps.txt
Comment (clason) [inline cmake.deps/deps.txt:63]:
```suggestion
WASMTIME_URL https://github.com/bytecodealliance/wasmtime/archive/v20.0.0.tar.gz
```

Like so -- that contains the  `wasm.h` and is how we track all other dependencies.

**Comment by dundargoc** [comment]
> I just checked; [bytecodealliance/wasmtime@`v20.0.0`.tar.gz](https://github.com/bytecodealliance/wasmtime/archive/v20.0.0.tar.gz) _does_ contain [`wasm.h`](https://github.com/bytecodealliance/wasmtime/blob/v20.0.0/crates/c-api/include/wasm.h?rgh-link-date=2024-04-23T10%3A56%3A20Z)? (release tag != release)

The previous releases didn't, so thanks for the gaslight wasmtime.

**Comment by clason** [comment]
That's right; they switched from a submodule to vendoring it for v20.0.0. Sorry, I should have checked older versions as well.

File: cmake.deps/deps.txt
Comment (clason) [inline cmake.deps/deps.txt:63]: probably want to update the SHA as well ;)

File: cmake.deps/deps.txt
Comment (dundargoc) [reply cmake.deps/deps.txt]: This has been fixed after the v20.0.0 release. See https://github.com/neovim/neovim/pull/28415#issuecomment-2072017990

File: runtime/doc/treesitter.txt
Comment (justinmk) [inline runtime/doc/treesitter.txt:41]:
```suggestion
EXPERIMENTAL: If Nvim is built with `ENABLE_WASMTIME`, then wasm parsers can also be
```

Comment (lewis6991) [reply runtime/doc/treesitter.txt:41]: Don't you mean `DOUBLY-EXPERIMENTAL` since all of treesitter is already experimental?

Comment (dundargoc) [reply runtime/doc/treesitter.txt:41]: 🧠

File: cmake.deps/cmake/BuildWasmtime.cmake
Comment (dundargoc) [reply cmake.deps/cmake/BuildWasmtime.cmake:5]: Nope, cmake is broken, I can't for the life of me get genex to work here. Always default to `--release` for the time being.

**Comment by dundargoc** [comment]
Maybe we should add a make variable that passes cmake args to both deps and neovim, `make CMAKE_EXTRA_FLAGS=-DENABLE_WASMTIME=ON DEPS_CMAKE_FLAGS=-DENABLE_WASMTIME=ON` is a bit much lol

File: cmake/Deps.cmake
Comment (dundargoc) [reply cmake/Deps.cmake]: Opted to only build treesitter with c11 when using wasmtime, this way we can postpone the discsussion of bumping to c11 to a later time.

**Comment by dundargoc** [comment]
TODO: make it work on windows

**Comment by lewis6991** [comment]
@clason @justinmk thoughts on merging now or just wait until the release?

**Comment by dundargoc** [comment]
Soft-dependent on https://github.com/bytecodealliance/wasmtime/pull/8490, aka would simplify some things but not a blocker.

File: .github/workflows/build.yml
Comment (justinmk) [inline .github/workflows/build.yml:32]: is this intentionally not ENABLE_WASMTIME ? same question for `cmake.deps/CMakePresets.json`

**Comment by justinmk** [comment]
> thoughts on merging now or just wait until the release?

No objection, because this doesn't introduce surface area that would be hard to change. If it causes any problems with the 0.10 release we can always hotfix stuff. That's different than e.g. a new API that would then need a deprecation cycle.

File: .github/workflows/build.yml
Comment (dundargoc) [reply .github/workflows/build.yml:32]: It's wrong. Thanks for the heads up.

File: cmake.deps/deps.txt
Comment (dundargoc) [inline cmake.deps/deps.txt:63]: Just use commit on my fork for now until https://github.com/bytecodealliance/wasmtime/pull/8496 is merged or a new version is released

Comment (justinmk) [reply cmake.deps/deps.txt:63]: supply chain

Comment (dundargoc) [reply cmake.deps/deps.txt:63]: Huh?

Comment (clason) [reply cmake.deps/deps.txt:63]: @dundargoc wasmtime 20.0.1 is released with your fix now.

**Comment by dundargoc** [comment]
This works on windows release-builds. The debug build fails with linker warnings. It would be good to get that to work, but not a blocker IMO. We can ship it as is for the time being when we can.

**Comment by clason** [comment]
Preliminary testing indicates that building nvim (and tree-sitter) with wasmtime has a significant performance impact, even without using wasm parsers: startuptime increases between 50% and 200% (depending on config and how you measure); the cost is somewhere in the Lua API stack.

(Comparing `--clean` shows little to no difference, so this is not related to the binary size increasing by ~20Mb due to the static linkage of wasmtime.)

**Comment by lewis6991** [comment]
If `--clean` shows no difference then I find that a bit sus as no extra code is called in `treesitter.c`.

**Comment by clason** [comment]
It _is_ sus, which is why I report it here as preliminary information. Obviously we need to profile where the increase comes from; it's entirely possible that it's from tree-sitter itself if built with wasmtime enabled.

**Comment by lewis6991** [comment]
`tslua_free` should only be needed for Asan.

**Comment by clason** [comment]
Yeah, it's not that (unsurprisingly); it was just an example of extra code in this PR. I have a hard time tracing the `--startuptime` differences to a specific point in my config.

**Comment by lewis6991** [comment]
You could try reverting all the C changes to see if the issue persists? That would rule out whether it's our code or not.

**Comment by clason** [comment]
That makes no difference. But it's weird that the startup time for (nearly) _all_ plugins increased across the board, including for `gitsigns` which has no treesitter interaction. Maybe the issue is a build flag for one of the other dependencies, or an optimization that fails with wasmtime statically linked? (Grasping at straws here, obviously.)

**Comment by clason** [comment]
After more voodoo debugging, here's a few findings:
1. performance hit seems to scale with binary size
2. using (full) `lto = true` drops the size by ~10 Mb and reduces the impact to ~10ms (from ~35ms)
3. nothing else (`strip  = true`, `opt-level = 's'`) seems to make a difference.

This is on macOS, by the way; also PUC Lua (or `jit.off`) may behave differently.

**Comment by lewis6991** [comment]
LTO generally makes a big difference on macos anyway so that makes sense. Maybe the change in binary size is causing the CPU caches to not work as they did before, and LTO just mitigates that.

This is where statically linking everything is beginning to show its costs. Could we maybe somehow link wasmtime dynamically (as a shared lib), but by using a shared lib we build? A sort of half way house between static and dynamic linking? Or this that a fake thing I just made up? @dundargoc

**Comment by clason** [comment]
That was my idea as well (it's what `/usr/local/lib/nvim/` is for, anyway), but @dundargoc mentioned that this is (borderline) impossible?

**Comment by lewis6991** [comment]
Maybe we just accept the hit and rely on LTO (and potentially any other magic flags).

Having wasm available opens some pretty big doors in general if we still want to add support for wasm plugins.

**Comment by clason** [comment]
> Maybe we just accept the hit and rely on LTO (and potentially any other magic flags).

Even with LTO (which increases build time by a factor of two, and requires hackery or forkery), I don't think it's acceptable to just shrug our shoulders and enable it by default. I note again that this is all without touching a single tree-sitter parser, let alone spooling up a wasmtime engine! An optional feature that negatively affects _all_ operations across the board is not a good trade-off. (And I'm saying that as the person who is most aggressively pushing for wasm parsers...)

**Comment by dundargoc** [comment]
> Could we maybe somehow link wasmtime dynamically (as a shared lib)

Yes, that should be possible.

> but by using a shared lib we build and ship with nvim (nvim-wasmtime)? A sort of half way house between static and dynamic linking?

We can build wasmtime as a shared library. The problem comes from that we can't use shared libraries for our releases, meaning that we won't be able to use wasmtime on our releases. Unsure if that's an acceptable trade, but I don't mind testing it out.

**Comment by dundargoc** [comment]
I've added some different flags and I don't notice a difference now. Please test again.

**Comment by lewis6991** [comment]
> Even with LTO (which increases build time by a factor of two, and requires hackery or forkery to even enable), I don't think it's acceptable to just shrug our shoulders and enable it by default.

The build time and binary size was mostly a given, no? I began this PR with that understanding.

The load time is the main concern for me, but ~10ms seems shruggable given the context. Especially if that means long term we can avoid users loading binary files, which I see as a huge security risk, amongst all the other issues.

The hackery and forkery should be temporary?

Just to note, I'm not aiming to merge this for 0.10, in case that wasn't understood. My mindset for this is for the longer term.

**Comment by clason** [comment]
My mindset as well. But it's _not_ just about load time; it looks to impact LuaJIT across the board (since the time difference scales strongly with complexity of the config). It's ok if the cost is associated with use, but that is very much not the case here.

> The build time and binary size was mostly a given, no? I began this PR with that understanding.

Yes, but we can (and should) still try to minimize that, given that they're quite egregious (increase by a factor of ~4).

> The hackery and forkery should be temporary?

That depends on whether wasmtime is willing to enable (much) more aggressive optimizations by default. I'm a bit skeptical about this (but surprises happen).

**Comment by clason** [comment]
But @dundargoc's forkery and hackery indeed closes that gap, so if we can get the necessary options (which we can probably reduce to two: `lto = true` and `code-points = 1`) into upstream (possibly as an additional profile we can just select from Cmake), then I think we're good.

File: cmake.deps/deps.txt
Comment (clason) [inline cmake.deps/deps.txt:63]: wrong org

**Comment by clason** [comment]
Update: it seems only macOS is (significantly) affected by linking wasmtime, not Linux, so that points to a linker issue with clang. (Would be interesting to check Windows, though.)

**Comment by dundargoc** [comment]
This is done. There are some followup work to be done such as making wasmtime work on debug windows as well as making the importing of the library more robust, but I will punt these tasks to a future date.

**Review by dundargoc — APPROVED** [review]

File: runtime/lua/vim/treesitter/language.lua
Comment (justinmk) [inline runtime/lua/vim/treesitter/language.lua:112]: should this be case-sensitive?

File: src/nvim/lua/treesitter.c
Comment (justinmk) [inline src/nvim/lua/treesitter.c:96]: should this match the convention of the existing `os_strerror`, `uv_strerror` ?
```suggestion
static const char *wasm_strerr(TSWasmErrorKind werr)
```

File: src/nvim/lua/treesitter.c
Comment (justinmk) [inline src/nvim/lua/treesitter.c:189]: should this mention `path` like the one below?
```suggestion
    luaL_error(L, "Failed to load WASM parser %s: (%s) %s", path, wasmerr_to_str(werr.kind), werr.message);
```

File: test/functional/treesitter/language_spec.lua
Comment (lewis6991) [reply test/functional/treesitter/language_spec.lua:65]: Lua doesn't have ternary, it has `cond and a or b`, however this shouldn't be used if `a` or `b` are booleans since `cond and false or b` -> `b` (`cond` is ignored).

In this case we could write `keys[k] = type(v) ~= 'boolean' or v`, but I think it does obscure the intention of the code.

Comment (justinmk) [reply test/functional/treesitter/language_spec.lua:65]: > Lua doesn't have ternary, it has `cond and a or b`

do i have to start typing that out or can I just say "ternary" :)

Comment (clason) [reply test/functional/treesitter/language_spec.lua:65]: The point is not the terminology; it's the fact that in Lua it's **not** equivalent to a conditional in cases involving booleans. This one of the major footguns we need to guard against.

Comment (lewis6991) [reply test/functional/treesitter/language_spec.lua:65]: The distinction is quite important and I'm not sure if you knew this. It has been the source of many bugs in core and plugins, including my own.

Comment (ofseed) [reply test/functional/treesitter/language_spec.lua:65]: I have proved with personal experience that we should avoid using ternary operations in Lua casually

https://github.com/neovim/neovim/pull/28627

**Comment by clason** [comment]
Status: Waiting on a tree-sitter release with wasmtime 22.0.0 (ETA: end of June/early July).

Reason: wasmtime 22.0.0 reworked their C API build to (hopefully) yield a lighter build since a) more symbols are hidden by a new header and b) optional features can be disabled

Reference:

* https://github.com/bytecodealliance/wasmtime/blob/fa84dd83ba0f5c45bf0174bb5f988c215d848cfe/crates/c-api/CMakeLists.txt#L44-L58 for the available feature flags
* https://lib.rs/crates/wasmtime/features for the description (everything can go except `cranelift`?)

**Comment by clason** [comment]
Update: That works (including actually loading a wasm parser), and cuts down wasmtime build by about half (60s -> 160s full nvim release build on my potato Intel mac). Binary size goes from 6.7 Mbyte to 12 Mbyte, startup from 30ms -> 30.5ms.

We could now even drop the full LTO from the build (60s -> 107s, 6.7 Mbyte -> 15 Mbyte, 30ms -> 31.5ms startup), but why say no to free performance? ;)

File: cmake.deps/cmake/BuildWasmtime.cmake
Comment (lewis6991) [inline cmake.deps/cmake/BuildWasmtime.cmake:9]: Can we document why we want these?

Comment (clason) [reply cmake.deps/cmake/BuildWasmtime.cmake:9]: To make wasmtime go *brrrrr*?

Comment (clason) [reply cmake.deps/cmake/BuildWasmtime.cmake:9]: * `FASTEST_RUNTIME` is basically `Release` on steroids (full lto)
* `DISABLE_ALL_FEATURES` does what it says on the tin -- a lot of optional stuff we don't need, except...
* `CRANELIFT`, which is the low-level code generator (required to compile wasm to platform code; otherwise no worky)

File: cmake.deps/cmake/BuildWasmtime.cmake
Comment (lewis6991) [inline cmake.deps/cmake/BuildWasmtime.cmake:1]: 🚀

File: cmake.deps/cmake/TreesitterCMakeLists.txt
Comment (clason) [inline cmake.deps/cmake/TreesitterCMakeLists.txt:8]: For the record, this is only needed to pick up https://github.com/tree-sitter/tree-sitter/blob/master/lib/src/wasm/stdlib-symbols.txt (list of exported symbols that are safe in wasm) during build.

Tree-sitter's own `build.rs` simply [copies that file](https://github.com/tree-sitter/tree-sitter/blob/3095fbe07bbcc605910b7ad8bbd6313e0a07e678/lib/binding_rust/build.rs#L9-L13). Would this be a (better) alternative? @dundargoc

**Comment by clason** [comment]
Rebased on master, but crashes (now?) when trying to edit a buffer with a tree-sitter wasm parser (tested with `vimdoc.wasm` loaded manually via `vim.treesitter.language.add()` on `treesitter.txt`; just make small edits and undo/redo).

**Comment by lewis6991** [comment]
Stack trace?

**Comment by clason** [comment]
No stack trace (macOS beta), not sure where the crash is.

File: cmake.deps/cmake/TreesitterCMakeLists.txt
Comment (dundargoc) [reply cmake.deps/cmake/TreesitterCMakeLists.txt:8]: That's basically what this does, "including" a directory here just means the directory will be copied during installation.

Comment (clason) [reply cmake.deps/cmake/TreesitterCMakeLists.txt:8]: My question was whether it would make sense to only copy the _file_, not the whole directory?

Comment (dundargoc) [reply cmake.deps/cmake/TreesitterCMakeLists.txt:8]: I don't think that's possible. I can't come up with a way at the top of my head.

**Comment by clason** [comment]
Note: Investigate how wasm parsers and native parsers for the same language interact (i.e., trying to (un)load a wasm parser after loading a native or vice versa).

**Comment by clason** [comment]
Note 2: Look at using [sccache](https://github.com/mozilla/sccache#usage) instead of `ccache` to speed up (re)building.

**Comment by dundargoc** [comment]
Idk, I didn't have much success with sccache with rust tbh. Maybe I didn't turn the correct knobs but I suspect it's just not very effective.

**Comment by clason** [comment]
I know, but maybe they fixed some stuff (by the time this will be enabled by default).

**Comment by lewis6991** [comment]
Is there a status update we can add to the OP?

**Comment by justinmk** [comment]
> trying to (un)load a wasm parser after loading a native or vice versa).

Does that mean it's possible to actually unload parsers now? I thought that wasn't really supported yet.

File: src/nvim/lua/treesitter.c
Comment (justinmk) [reply src/nvim/lua/treesitter.c:96]: @lewis6991

**Comment by ofseed** [comment]
Since there is wasm parser support already, will there be support for downloading these parsers by the neovim or nvim-treesitter organization? There's a version lock file in the nvim-treesitter repo, I think it would be great if nvim-treesitter had a CI to build them and we can download it from the release.

**Pull request merged**
